clients/drcachesim/tests/raw2trace_unit_tests.cpp

```diff
@@ -929,21 +936,127 @@ test_duplicate_syscalls(void *drcontext)
-        // No duplicate sys instr.
-        // We keep the extraneous timestamp+cpu markers above.
-        // Prev block ends.
-        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP) &&
+        // No duplicate sys instr, and the following timestamp==3 and syscall marker
+        // are removed.
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, 4) &&
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID) &&
+        // The move2 instr.
+        check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+        check_entry(entries, idx, TRACE_TYPE_INSTR, -1) &&
+        check_entry(entries, idx, TRACE_TYPE_THREAD_EXIT, -1) &&
+        check_entry(entries, idx, TRACE_TYPE_FOOTER, -1));
+}
+
+bool
+test_false_syscalls(void *drcontext)
+{
+    std::cerr << "\n===============\nTesting false syscalls\n";
+    // Our synthetic test first constructs a list of instructions to be encoded into
+    // a buffer for decoding by raw2trace.
+    instrlist_t *ilist = instrlist_create(drcontext);
+    // raw2trace doesn't like offsets of 0 so we shift with a nop.
+    instr_t *nop = XINST_CREATE_nop(drcontext);
+    instr_t *move1 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    // XXX: Adding an XINST_CREATE_syscall macro will simplify this but there are
+    // complexities (xref create_syscall_instr()).
+#ifdef X86
+    instr_t *sys = INSTR_CREATE_syscall(drcontext);
+#elif defined(AARCHXX)
+    instr_t *sys = INSTR_CREATE_svc(drcontext, opnd_create_immed_int((sbyte)0x0, OPSZ_1));
+#elif defined(RISCV64)
+    instr_t *sys = INSTR_CREATE_ecall(drcontext);
+#else
+#    error Unsupported architecture.
+#endif
+    instr_t *move2 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG2), opnd_create_reg(REG1));
+    instrlist_append(ilist, nop);
+    instrlist_append(ilist, move1);
+    instrlist_append(ilist, sys);
+    instrlist_append(ilist, move2);
+    size_t offs_nop = 0;
+    size_t offs_move1 = offs_nop + instr_length(drcontext, nop);
+    size_t offs_sys = offs_move1 + instr_length(drcontext, move1);
+    size_t offs_move2 = offs_sys + instr_length(drcontext, sys);
+
+    // Now we synthesize our raw trace itself, including a valid header sequence.
+    static constexpr int SYSCALL_NUM = 42; // Doesn't really matter.
+    std::vector<offline_entry_t> raw;
+    raw.push_back(make_header());
+    raw.push_back(make_tid());
+    raw.push_back(make_pid());
+    raw.push_back(make_line_size());
+    raw.push_back(make_timestamp(1));
+    raw.push_back(make_core());
+    raw.push_back(make_block(offs_move1, 2));
+    // There is no syscall marker here, so the syscall should be removed.
+    raw.push_back(make_timestamp(2));
+    raw.push_back(make_core());
+    // Repeat the syscall but with a marker this time.
+    // Start with the move to not trigger dup syscall removal.
```

Comment:
"Start with the move": is this the scenario where the block-interrupting-signal interrupts the block after we emit the pc entry but before any instr was executed? So when we restart from the block, we emit the pc entry again, so we see the `(offs_move1,2)` again?

Reply:
If yes: then the correct output is that the first instance of both the move1 and syscall must be removed. Can we add a `TODO i#` by the expected output below?

Reply:
See the prior thread on this: that line "Start with the move to not trigger dup syscall remove" is incorrect and is from the early PR where dup syscall remove went first.  That line is now removed and the block on the line below starts with the syscall itself.